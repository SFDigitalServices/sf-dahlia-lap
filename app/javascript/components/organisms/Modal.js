import React from 'react'

import ReactModal from 'react-modal'

import Alert from '../organisms/Alert'

const styleTypes = {
  small: {
    overlay: {
      backgroundColor: 'rgba(0,0,0,0.3)'
    },
    content: {
      opacity: '1',
      border: '0',
      borderRadius: '4px',
      bottom: 'auto',
      minHeight: '10rem',
      left: '50%',
      top: '50%',
      padding: '0px',
      position: 'fixed',
      right: 'auto',
      overflow: 'visible',
      transform: 'translate(-50%,-50%)',
      width: '600px'
    }
  }
}

const Modal = ({ children, isOpen, handleClose, styleType }) => (
  <ReactModal
    isOpen={isOpen}
    style={styleTypes[styleType || 'small']}
    onRequestClose={handleClose}
    shouldCloseOnOverlayClick
    ariaHideApp={process.env.NODE_ENV !== 'test'}
  >
    {children}
  </ReactModal>
)

Modal.Body = ({ children, handleClose, hidden }) => (
  <div aria-labelledby='modalTitle' aria-hidden={hidden} role='dialog'>
    {children}
    <button
      className='button button-link close-reveal-modal'
      aria-label='Close modal'
      onClick={handleClose}
    >
      <span className='sr-only'>Close</span>
      <span className='ui-icon ui-medium i-primary'>
        <svg>
          <use xlinkHref='#i-close' />
        </svg>
      </span>
    </button>
  </div>
)

Modal.Header = ({ id, title }) => (
  <header id={id} className='modal-inner margin-top'>
    <h1 className='modal-title t-gamma no-margin'>{title}</h1>
  </header>
)

Modal.Content = ({ children }) => <section className='modal-inner'>{children}</section>

Modal.Footer = ({ children }) => (
  <footer className='modal-footer bg-dust'>
    <div className='modal-button-group row'>{children}</div>
  </footer>
)

Modal.Alert = (props) => <Alert {...props} />

export default Modal
