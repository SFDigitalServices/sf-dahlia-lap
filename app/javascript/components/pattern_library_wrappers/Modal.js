import React from 'react'
import ReactDOM from 'react-dom';
import ReactModal from 'react-modal';

import SimpleModal from 'components/organisms/SimpleModal'

class ModalWrapper extends React.Component {
  constructor() {
    super()

    this.state = {
      statusModalIsOpen: false,
      dangerModalIsOpen: false
    };

    this.openStatusModal = this.openStatusModal.bind(this)
    // this.afterOpenModal = this.afterOpenModal.bind(this)
    this.closeStatusModal = this.closeStatusModal.bind(this)

    this.openDangerModal = this.openDangerModal.bind(this)
    this.closeDangerModal = this.closeDangerModal.bind(this)
  }

  openStatusModal() {
    this.setState({statusModalIsOpen: true});
  }

  closeStatusModal() {
    this.setState({statusModalIsOpen: false});
  }

  openDangerModal() {
    this.setState({dangerModalIsOpen: true});
  }

  closeDangerModal() {
    this.setState({dangerModalIsOpen: false});
  }

  render() {
    return (
      <div>
        <button onClick={this.openStatusModal}>Open Update Status Modal</button>
        <SimpleModal
          header='Update Status'
          primary='update'
          secondary='cancel'
          isOpen={this.state.statusModalIsOpen}
          onCloseClick={() => this.closeStatusModal()}
          onPrimaryClick={() => this.closeStatusModal()}
          onSecondaryClick={() => this.closeStatusModal()}
          type='status'>
          <div>content</div>
        </SimpleModal>
        <br/>

        <button onClick={this.openDangerModal}>Open Danger Modal</button>
        <SimpleModal
          header='Remove loterry prefrence?'
          primary='remove'
          secondary='cancel'
          isOpen={this.state.dangerModalIsOpen}
          onCloseClick={() => this.closeDangerModal()}
          onPrimaryClick={() => this.closeDangerModal()}
          onSecondaryClick={() => this.closeDangerModal()}
          type='alert'
          alertBox="This change will affect this application's preferences"
          alertNotice={ {
            title: 'This application woudl no longer be elegible for Live Work Preference',
            content: 'Note, you will have the opportunity to grant another household member this preference'
          }}>
          <div>content</div>
        </SimpleModal>

        <br/>
        <button onClick={this.openDangerModal}>Open Invert Modal</button>
        <SimpleModal
          header='Remove loterry prefrence?'
          primary='remove'
          secondary='cancel'
          isOpen={this.state.dangerModalIsOpen}
          onCloseClick={() => this.closeDangerModal()}
          onPrimaryClick={() => this.closeDangerModal()}
          onSecondaryClick={() => this.closeDangerModal()}
          type='alert'
          alertBox="This change will affect this application's preferences"
          alertNotice={ {
            title: 'This application woudl no longer be elegible for Live Work Preference',
            content: 'Note, you will have the opportunity to grant another household member this preference'
          }}>
          <div>content</div>
        </SimpleModal>
      </div>
    )
  }
}

export default ModalWrapper
