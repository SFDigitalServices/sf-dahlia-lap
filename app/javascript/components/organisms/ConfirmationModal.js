import React from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'

import Modal from './Modal'

const ConfirmationModal = ({
  isOpen = false,
  onCloseClick = () => {},
  onPrimaryClick = () => {},
  onSecondaryClick = () => {},
  primaryText,
  primaryButtonDestination = null,
  primaryButtonIsAlert = false,
  secondaryText,
  subtitle,
  title,
  titleId = 'confirmation-modal-header'
}) => {
  const buttonClasses = classNames(
    'button',
    primaryButtonIsAlert ? 'alert' : 'primary'
  )

  return (
    <Modal isOpen={isOpen} handleClose={onCloseClick}>
      <Modal.Body hidden={isOpen} handleClose={onCloseClick}>
        <Modal.Header id={titleId} title={title} />
        <Modal.Content>
          <p>{subtitle}</p>
        </Modal.Content>
        <Modal.Footer>
          <div className='modal-button_item modal-button_secondary'>
            <a
              className={buttonClasses}
              href={primaryButtonDestination || '#'}
              type='button'
              onClick={onPrimaryClick}
            >
              {primaryText}
            </a>
          </div>
          <div className='modal-button_item modal-button_secondary'>
            <button className='button no-border' onClick={onSecondaryClick} type='button'>
              {secondaryText}
            </button>
          </div>
        </Modal.Footer>
      </Modal.Body>
    </Modal>
  )
}

ConfirmationModal.propTypes = {
  isOpen: PropTypes.bool,
  onCloseClick: PropTypes.func,
  onPrimaryClick: PropTypes.func,
  onSecondaryClick: PropTypes.func,
  primaryButtonDestination: PropTypes.string,
  primaryButtonIsAlert: PropTypes.bool,
  primaryText: PropTypes.string,
  secondaryText: PropTypes.string,
  subtitle: PropTypes.string,
  title: PropTypes.string,
  titleId: PropTypes.string
}

export default ConfirmationModal
