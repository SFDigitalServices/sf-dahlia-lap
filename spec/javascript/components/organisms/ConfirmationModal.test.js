import React from 'react'
import { shallow } from 'enzyme'

import ConfirmationModal from '~/components/organisms/ConfirmationModal'
import Modal from '~/components/organisms/Modal'

import { findWithText } from '../../testUtils/wrapperUtil'

const PRIMARY_TEXT = 'Primary Text'
const SECONDARY_TEXT = 'Secondary Text'
const TITLE = 'Title'
const SUBTITLE = 'Subtitle'

const ON_CLOSE = jest.fn()
const ON_PRIMARY_CLICK = jest.fn()
const ON_SECONDARY_CLICK = jest.fn()

const getWrapper = (propOverrides = {}) => shallow(
  <ConfirmationModal
    onPrimaryClick={ON_PRIMARY_CLICK}
    onSecondaryClick={ON_SECONDARY_CLICK}
    onCloseClick={ON_CLOSE}
    primaryText={PRIMARY_TEXT}
    secondaryText={SECONDARY_TEXT}
    subtitle={SUBTITLE}
    title={TITLE}
    {...propOverrides}
  />
)

describe('ConfirmationModal', () => {
  describe('with default props', () => {
    let wrapper
    beforeEach(() => {
      wrapper = getWrapper()
    })

    test('should render the modal as closed', () => {
      expect(wrapper.find(Modal).prop('isOpen')).toBeFalsy()
      expect(wrapper.find(Modal.Body).prop('hidden')).toBeFalsy()
    })

    test('should render the title and subtitle', () => {
      expect(wrapper.find(Modal.Header).prop('title')).toEqual(TITLE)
      expect(wrapper.find(Modal.Content).dive().text()).toEqual(SUBTITLE)
    })

    test('should render with the default header id', () => {
      expect(wrapper.find(Modal.Header).prop('id')).toEqual('confirmation-modal-header')
    })

    test('should render the primary button as a link with href=#', () => {
      const primaryButtonWrapper = findWithText(wrapper, 'a', PRIMARY_TEXT)
      expect(primaryButtonWrapper).toHaveLength(1)
      expect(primaryButtonWrapper.prop('href')).toEqual('#')
    })

    test('should render the primary button with primary style', () => {
      const primaryWrapper = findWithText(wrapper, 'a', PRIMARY_TEXT)
      expect(primaryWrapper.prop('className').includes('primary')).toBeTruthy()
    })

    test('should render the secondary button', () => {
      expect(findWithText(wrapper, 'button', SECONDARY_TEXT)).toHaveLength(1)
    })

    test('should render the secondary button', () => {
      expect(findWithText(wrapper, 'button', SECONDARY_TEXT)).toHaveLength(1)
    })

    test('should trigger the onClose listener when close is clicked', () => {
      wrapper.find(Modal).prop('handleClose')()

      expect(ON_CLOSE.mock.calls.length).toEqual(1)
      expect(ON_PRIMARY_CLICK.mock.calls.length).toEqual(0)
      expect(ON_SECONDARY_CLICK.mock.calls.length).toEqual(0)
    })

    test('should trigger the onPrimaryClick listener when primary button is clicked', () => {
      findWithText(wrapper, 'a', PRIMARY_TEXT).simulate('click')

      expect(ON_CLOSE.mock.calls.length).toEqual(0)
      expect(ON_PRIMARY_CLICK.mock.calls.length).toEqual(1)
      expect(ON_SECONDARY_CLICK.mock.calls.length).toEqual(0)
    })

    test('should trigger the onSecondaryClick listener when secondary button is clicked', () => {
      findWithText(wrapper, 'button', SECONDARY_TEXT).simulate('click')

      expect(ON_CLOSE.mock.calls.length).toEqual(0)
      expect(ON_PRIMARY_CLICK.mock.calls.length).toEqual(0)
      expect(ON_SECONDARY_CLICK.mock.calls.length).toEqual(1)
    })
  })

  describe('with custom titleId', () => {
    const CUSTOM_TITLE_ID = 'custom-title-id'
    let wrapper
    beforeEach(() => {
      wrapper = getWrapper({ titleId: CUSTOM_TITLE_ID })
    })

    test('should render with the custom header id', () => {
      expect(wrapper.find(Modal.Header).prop('id')).toEqual(CUSTOM_TITLE_ID)
    })
  })

  describe('when isOpen is true', () => {
    let wrapper
    beforeEach(() => {
      wrapper = getWrapper({ isOpen: true })
    })

    test('should render the modal as open', () => {
      expect(wrapper.find(Modal).prop('isOpen')).toBeTruthy()
      expect(wrapper.find(Modal.Body).prop('hidden')).toBeTruthy()
    })
  })

  describe('when primaryButtonIsAlert is true', () => {
    let wrapper
    beforeEach(() => {
      wrapper = getWrapper({ primaryButtonIsAlert: true })
    })

    test('should render the primary button with alert styling', () => {
      const primaryWrapper = findWithText(wrapper, 'a', PRIMARY_TEXT)
      expect(primaryWrapper.prop('className').includes('alert')).toBeTruthy()
    })
  })

  describe('when primaryButtonDestination is provided', () => {
    const DESTINATION = '/destinationPath'
    let wrapper
    beforeEach(() => {
      wrapper = getWrapper({ primaryButtonDestination: DESTINATION })
    })

    test('should render the primary button with the correct href', () => {
      const primaryWrapper = findWithText(wrapper, 'a', PRIMARY_TEXT)
      expect(primaryWrapper.prop('href')).toEqual(DESTINATION)
    })
  })
})
