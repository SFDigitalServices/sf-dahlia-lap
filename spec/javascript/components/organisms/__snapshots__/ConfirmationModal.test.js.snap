// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmationModal should render the modal as closed 1`] = `<DocumentFragment />`;
